iot/mqtt_example/cloudiot_mqtt_example_nodejs.js

```diff
@@ -222,88 +352,470 @@ function publishAsync(messagesSent, numMessages) {
+    if (backoffTime >= MAXIMUM_BACKOFF_TIME) {
+      console.log('Backoff time is too high. Giving up.');
+    }
+    if (messagesSent >= numMessages) {
+      detachDevice(deviceId, client);
+    }
+    console.log('Closing connection to MQTT. Goodbye!');
+    client.end();
+    publishChainInProgress = false;
+    return;
+  }
 
-// Once all of the messages have been published, the connection to Google Cloud
-// IoT will be closed and the process will exit. See the publishAsync method.
-// [END iot_mqtt_run]
+  // Publish and schedule the next publish.
+  publishChainInProgress = true;
+  var publishDelayMs = 0;
+  if (shouldBackoff) {
+    publishDelayMs = 1000 * (backoffTime + Math.random());
+    backoffTime *= 2;
+    console.log(`Backing off for ${publishDelayMs}ms before publishing.`);
+  }
+  let mqttTopic = `/devices/${gatewayId}/state`;
+  let payload = `${registryId}/${gatewayId}-connected-${new Date().getTime()}`;
+  console.log(`Publishing message ${messagesSent}/${numMessages}`);
+  if (messagesSent > 0) {
+    mqttTopic = `/devices/${deviceId}/state`;
+    payload = `${registryId}/${deviceId}-payload-${messagesSent}`;
+  }
+
+  setTimeout(function() {
+    // Publish "payload" to the MQTT topic. qos=1 means at least once delivery.
+    // Cloud IoT Core also supports qos=0 for at most once delivery.
+    console.log(`Publishing message: ${payload} to ${mqttTopic}`);
+    client.publish(mqttTopic, payload, {qos: 1}, function(err) {
+      if (!err) {
+        shouldBackoff = false;
+        backoffTime = MINIMUM_BACKOFF_TIME;
+      }
+    });
+
+    var schedulePublishDelayMs = 5000; // messageType === 'events' ? 1000 : 2000;
+    setTimeout(function() {
+      let secsFromIssue = parseInt(Date.now() / 1000) - iatTime;
+      if (secsFromIssue > tokenExpMins * 60) {
+        iatTime = parseInt(Date.now() / 1000);
+        console.log(`\tRefreshing token after ${secsFromIssue} seconds.`);
+
+        client.end();
+        connectionArgs.password = createJwt(
+          projectId,
+          privateKeyFile,
+          algorithm
+        );
+        client = mqtt.connect(connectionArgs);
+      }
+      publishAsyncGateways(
+        client,
+        iatTime,
+        tokenExpMins,
+        messagesSent + 1,
+        numMessages,
+        registryId,
+        deviceId,
+        gatewayId
+      );
+    }, schedulePublishDelayMs);
+  }, publishDelayMs);
+}
+
+// Sends data from a gateway on behalf of a device that is bound to that gateway.
+function sendDataFromBoundDevice(
+  gatewayId,
+  deviceId,
+  registryId,
+  projectId,
+  region,
+  algorithm,
+  privateKeyFile,
+  mqttBridgeHostname,
+  mqttBridgePort,
+  numMessages,
+  tokenExpMins
+) {
+  // [START iot_send_data_from_bound_device]
+  // const parentName = `projects/${projectId}/locations/${region}`;
+  // const registryName = `${parentName}/registries/${registryId}`;
+  // const binaryData = Buffer.from(data).toString('base64');
```

Comment:
Copy/paste error maybe?  In the NodeJS examples we show the variables with examples so that they appear in the docs.

I'd put it maybe as:

    // const gatewayId =  `mygateway`;
    // const deviceId = `myDevice`;
    // const registryId = `myRegistry`;
    // const region = `us-central1`;
    // const algorithm = `RS256`;
    // const privateKeyFile = `./rsa_private.pem`;
    // const mqttBridgeHostname = `mqtt.googleapis.com`;
    // const mqttBridgePort = 8883;
    // const numMessages = 5;
    // const tokenExpMins = 60;